### ISSUE TITLE
kubectl run --dry-run raises error

### ISSUE THREAD

**Issue by abstrctn** [kubernetes/kubectl#45]
**What keywords did you search in Kubernetes issues before filing this one?**: There were several issues in github.com/kubernetes/kubernetes, chiefly this one that suggested creating an issue in this repo instead, which I'm doing now since nothing seems to exist in this repo: https://github.com/kubernetes/kubernetes/issues/47180

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Bug report

**Kubernetes version** (use `kubectl version`): kubectl 1.7.0, server 1.7.1-gke.0

**Environment**:
- **Cloud provider or hardware configuration**: GKE 1.7.1

**What happened**:
Running `kubectl run nginx --image=nginx --dry-run`, as suggested in the [docs](https://kubernetes.io/docs/user-guide/kubectl/v1.7/#run), raises an error:

```
error: unknown type *v1beta1.Deployment, expected unstructured in map[reflect.Type]*printers.handlerEntry{(*reflect.rtype)(0x2c435c0):(*printers.handlerEntry)(0xc4203f9b80), (*reflect.rtype)(0x2c442e0):(*printers.handlerEntry)(0xc4206341e0), (*reflect.rtype)(0x2b3ce20):(*printers.handlerEntry)(0xc420635400), (*reflect.rtype)(0x2c40140):(*printers.handlerEntry)(0xc4203f8280), (*reflect.rtype)(0x2c3df20):(*printers.handlerEntry)(0xc4203f8960), (*reflect.rtype)(0x2c403e0):(*printers.handlerEntry)(0xc4203f90e0), (*reflect.rtype)(0x2c3e9a0):(*printers.handlerEntry)(0xc4203f93b0), (*reflect.rtype)(0x2c43da0):(*printers.handlerEntry)(0xc4203f9680), (*reflect.rtype)(0x2b3dc20):(*printers.handlerEntry)(0xc420635680), (*reflect.rtype)(0x2b3d360):(*printers.handlerEntry)(0xc4204fbbd0), (*reflect.rtype)(0x2b3cf00):(*printers.handlerEntry)(0xc4204fbd10), (*reflect.rtype)(0x2b3c480):(*printers.handlerEntry)(0xc4203f8320), (*reflect.rtype)(0x2b3caa0):(*printers.handlerEntry)(0xc4203f99a0), (*reflect.rtype)(0x2b3cfe0):(*printers.handlerEntry)(0xc4203f9860), (*reflect.rtype)(0x2b3bbc0):(*printers.handlerEntry)(0xc4203f8e60), (*reflect.rtype)(0x2b3d440):(*printers.handlerEntry)(0xc4203f9d60), (*reflect.rtype)(0x2b3d600):(*printers.handlerEntry)(0xc4206343c0), (*reflect.rtype)(0x2c45540):(*printers.handlerEntry)(0xc4206355e0), (*reflect.rtype)(0x2c3f960):(*printers.handlerEntry)(0xc4204fb950), (*reflect.rtype)(0x2b3bca0):(*printers.handlerEntry)(0xc4203f88c0), (*reflect.rtype)(0x2b3b680):(*printers.handlerEntry)(0xc4203f95e0), (*reflect.rtype)(0x2b3b840):(*printers.handlerEntry)(0xc4203f86e0), (*reflect.rtype)(0x2c3fea0):(*printers.handlerEntry)(0xc4203f8fa0), (*reflect.rtype)(0x2b3be60):(*printers.handlerEntry)(0xc4203f94a0), (*reflect.rtype)(0x2c457e0):(*printers.handlerEntry)(0xc4206354a0), (*reflect.rtype)(0x2c3d9e0):(*printers.handlerEntry)(0xc4203f9a40), (*reflect.rtype)(0x2c44040):(*printers.handlerEntry)(0xc4203f9cc0), (*reflect.rtype)(0x2b3d980):(*printers.handlerEntry)(0xc420634dc0), (*reflect.rtype)(0x2b3cc60):(*printers.handlerEntry)(0xc4203f81e0), (*reflect.rtype)(0x2b3d0c0):(*printers.handlerEntry)(0xc4203f8460), (*reflect.rtype)(0x2b3c8e0):(*printers.handlerEntry)(0xc4203f85a0), (*reflect.rtype)(0x2c3e460):(*printers.handlerEntry)(0xc4203f8dc0), (*reflect.rtype)(0x2c3ec40):(*printers.handlerEntry)(0xc4203f9270), (*reflect.rtype)(0x2c40bc0):(*printers.handlerEntry)(0xc4206357c0), (*reflect.rtype)(0x2b3cd40):(*printers.handlerEntry)(0xc4204fbe50), (*reflect.rtype)(0x2c3dc80):(*printers.handlerEntry)(0xc4203f8640), (*reflect.rtype)(0x2b3d8a0):(*printers.handlerEntry)(0xc420635130), (*reflect.rtype)(0x2bedce0):(*printers.handlerEntry)(0xc420635720), (*reflect.rtype)(0x2c44ac0):(*printers.handlerEntry)(0xc420634b40), (*reflect.rtype)(0x2b3bf40):(*printers.handlerEntry)(0xc4204fb450), (*reflect.rtype)(0x2c3e700):(*printers.handlerEntry)(0xc4203f8780), (*reflect.rtype)(0x2b3b920):(*printers.handlerEntry)(0xc4203f8a00), (*reflect.rtype)(0x2b3c1e0):(*printers.handlerEntry)(0xc4203f8d20), (*reflect.rtype)(0x2c3d740):(*printers.handlerEntry)(0xc4203f9540), (*reflect.rtype)(0x2c420c0):(*printers.handlerEntry)(0xc4204fbf90), (*reflect.rtype)(0x2b3d520):(*printers.handlerEntry)(0xc4203f9720), (*reflect.rtype)(0x2b3b760):(*printers.handlerEntry)(0xc4203f9ae0), (*reflect.rtype)(0x2c3d200):(*printers.handlerEntry)(0xc4203f9e00), (*reflect.rtype)(0x2c452a0):(*printers.handlerEntry)(0xc4206348c0), (*reflect.rtype)(0x2b3d7c0):(*printers.handlerEntry)(0xc4204fb8b0), (*reflect.rtype)(0x2c43860):(*printers.handlerEntry)(0xc4204fba90), (*reflect.rtype)(0x2c42360):(*printers.handlerEntry)(0xc4204fbdb0), (*reflect.rtype)(0x2b3c2c0):(*printers.handlerEntry)(0xc4203f9040), (*reflect.rtype)(0x2c42de0):(*printers.handlerEntry)(0xc4203f97c0), (*reflect.rtype)(0x2b3dd00):(*printers.handlerEntry)(0xc420635540), (*reflect.rtype)(0x2c41b80):(*printers.handlerEntry)(0xc4203f9900), (*reflect.rtype)(0x2b3d1a0):(*printers.handlerEntry)(0xc4206340a0), (*reflect.rtype)(0x2c45000):(*printers.handlerEntry)(0xc420634500), (*reflect.rtype)(0x2c3f420):(*printers.handlerEntry)(0xc4204fb5e0), (*reflect.rtype)(0x2b3c020):(*printers.handlerEntry)(0xc4204fb6d0), (*reflect.rtype)(0x2c40e60):(*printers.handlerEntry)(0xc4203f8500), (*reflect.rtype)(0x2c3e1c0):(*printers.handlerEntry)(0xc4203f8aa0), (*reflect.rtype)(0x2c3fc00):(*printers.handlerEntry)(0xc4203f8c80), (*reflect.rtype)(0x2b3da60):(*printers.handlerEntry)(0xc420634a00), (*reflect.rtype)(0x2c44820):(*printers.handlerEntry)(0xc4204fb810), (*reflect.rtype)(0x2c42b40):(*printers.handlerEntry)(0xc4204fbc70), (*reflect.rtype)(0x2c43080):(*printers.handlerEntry)(0xc4203f83c0), (*reflect.rtype)(0x2b3c100):(*printers.handlerEntry)(0xc4204fb9f0), (*reflect.rtype)(0x2b3bd80):(*printers.handlerEntry)(0xc4203f9310), (*reflect.rtype)(0x2b3b220):(*printers.handlerEntry)(0xc4203f9ea0), (*reflect.rtype)(0x2c43320):(*printers.handlerEntry)(0xc4203f9f40), (*reflect.rtype)(0x2c44d60):(*printers.handlerEntry)(0xc420634fa0), (*reflect.rtype)(0x2b3db40):(*printers.handlerEntry)(0xc4206346e0), (*reflect.rtype)(0x2c428a0):(*printers.handlerEntry)(0xc420635360), (*reflect.rtype)(0x2c3eee0):(*printers.handlerEntry)(0xc4204fb4a0), (*reflect.rtype)(0x2b3ba00):(*printers.handlerEntry)(0xc4203f8b40), (*reflect.rtype)(0x2b3c3a0):(*printers.handlerEntry)(0xc4203f9180), (*reflect.rtype)(0x2b3d280):(*printers.handlerEntry)(0xc4203f9c20), (*reflect.rtype)(0x2b3c800):(*printers.handlerEntry)(0xc420635860)}
```

**What you expected to happen**:
It should print out the object that would be sent to the server.

**How to reproduce it** (as minimally and precisely as possible):
1. Create a new GKE cluster running version 1.7.1
2. Configure kubectl to use the new cluster as it's default context
3. Run: `kubectl run nginx --image=nginx --dry-run`

**Comment by pwittrock** [comment]
@mengqiy @shiywang Could you look into this?

**Comment by pwittrock** [comment]
Is this related to #48659

**Comment by mengqiy** [comment]
> Is this related to kubernetes/kubernetes#48659

I don't think so.

**Comment by pwittrock** [comment]
@abstrctn  Thanks for filing this issue.  I was able to reproduce, though I am curious about your use case.  Is it just to validate that the command line args are correct, or something else?

**Comment by zjj2wry** [comment]
i cannot reproduce this. maybe has fix ??
```
admindeMacBook-Pro-2:kubectl admin$ ./kubectl run nginx --image=nginx --dry-run
NAME      AGE
nginx     <unknown>
```

**Comment by abstrctn** [comment]
@pwittrock Sorry for my late response - I probably should have included my original goal in the issue so wouldn't forget it... It does seem to still be happening on my version of kubectl, though I know it's not the latest (1.7.3)

Using `-o yaml` or json should work fine though. Normally we use dry-run inside automation scripts to template out baseline resources, which we then modify using `jq` (in cases where `--overrides` is granular enough).

**Comment by apelisse** [comment]
> Normally we use dry-run inside automation scripts to template out baseline resources, which we then modify using `jq`

Don't you use that with `-o yaml/json` though?

**Comment by mengqiy** [comment]
I can reproduce this issue with 1.7.5 kubectl
But I found kubectl built from HEAD works fine.

**Comment by pwittrock** [comment]
Apparently fixed in 1.8
